src/Storages/StorageIceberg.cpp

```diff
@@ -0,0 +1,422 @@
+#include "config.h"
+#if USE_AWS_S3
+
+#    include <Storages/StorageIceberg.h>
+#    include <Common/logger_useful.h>
+
+#    include <Columns/ColumnString.h>
+#    include <Columns/ColumnTuple.h>
+#    include <Columns/IColumn.h>
+
+#    include <IO/ReadBufferFromS3.h>
+#    include <IO/ReadHelpers.h>
+#    include <IO/ReadSettings.h>
+#    include <IO/S3Common.h>
+
+#    include <Storages/ExternalDataSourceConfiguration.h>
+#    include <Storages/StorageFactory.h>
+#    include <Storages/checkAndGetLiteralArgument.h>
+
+#    include <Formats/FormatFactory.h>
+
+#    include <aws/core/auth/AWSCredentials.h>
+#    include <aws/s3/S3Client.h>
+#    include <aws/s3/model/ListObjectsV2Request.h>
+
+#    include <QueryPipeline/Pipe.h>
+
+#    include <fmt/format.h>
+#    include <fmt/ranges.h>
+#    include <ranges>
+
+#    include <Processors/Formats/Impl/AvroRowInputFormat.h>
+
+#    include <Poco/JSON/Array.h>
+#    include <Poco/JSON/Object.h>
+#    include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int S3_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int FILE_DOESNT_EXIST;
+    extern const int ILLEGAL_COLUMN;
+}
+
+IcebergMetaParser::IcebergMetaParser(const StorageS3::S3Configuration & configuration_, const String & table_path_, ContextPtr context_)
+    : base_configuration(configuration_), table_path(table_path_), context(context_)
+{
+}
+
+std::vector<String> IcebergMetaParser::getFiles() const
+{
+    auto metadata = getNewestMetaFile();
+    auto manifest_list = getManiFestList(metadata);
+
+    /// When table first created and does not have any data
+    if (manifest_list.empty())
+    {
+        return {};
+    }
+
+    auto manifest_files = getManifestFiles(manifest_list);
+    return getFilesForRead(manifest_files);
+}
+
+String IcebergMetaParser::getNewestMetaFile() const
+{
+    /// Iceberg stores all the metadata.json in metadata directory, and the
+    /// newest version has the max version name, so we should list all of them
+    /// then find the newest metadata.
+    std::vector<String> metadata_files;
+
+    const auto & client = base_configuration.client;
+
+    Aws::S3::Model::ListObjectsV2Request request;
+    Aws::S3::Model::ListObjectsV2Outcome outcome;
```

Comment:
let's make a function in s3Common like `listFiles` by some prefix and reuse it here